src/backend/distributed/commands/database.c

```diff
@@ -242,3 +302,193 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
+ * create/drop a database in an idempotent maner without
+ * transaction block restrictions.
+ */
+Datum
+citus_internal_database_command(PG_FUNCTION_ARGS)
+{
+	int saveNestLevel = NewGUCNestLevel();
+	text *commandText = PG_GETARG_TEXT_P(0);
+	char *command = text_to_cstring(commandText);
+	Node *parseTree = ParseTreeNode(command);
+
+	set_config_option("citus.enable_ddl_propagation", "off",
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+
+	set_config_option("citus.enable_create_database_propagation", "off",
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+
+	if (IsA(parseTree, CreatedbStmt))
+	{
+		CreatedbStmt *stmt = castNode(CreatedbStmt, parseTree);
+
+		bool missingOk = true;
+		Oid databaseOid = get_database_oid(stmt->dbname, missingOk);
+
+		if (!OidIsValid(databaseOid))
+		{
+			createdb(NULL, (CreatedbStmt *) parseTree);
+		}
+		else
+		{
+			/* TODO: check database properties */
+		}
+	}
+	else if (IsA(parseTree, DropdbStmt))
+	{
+		DropdbStmt *stmt = castNode(DropdbStmt, parseTree);
+
+		bool missingOk = false;
+		Oid databaseOid = get_database_oid(stmt->dbname, missingOk);
+
+		if (!OidIsValid(databaseOid))
+		{
+			/* already dropped? */
+		}
+		else
+		{
+			/* / * remove database from database shards * / */
+			/* DeleteDatabaseShardByDatabaseIdLocally(databaseOid); */
+
+			DropDatabase(NULL, (DropdbStmt *) parseTree);
+		}
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("unsupported command type %d", nodeTag(parseTree))));
+	}
+
+	AtEOXact_GUC(true, saveNestLevel);
+
+	PG_RETURN_VOID();
+}
+
+
+List *
+PreprocessDropDatabaseStmt(Node *node, const char *queryString,
+						   ProcessUtilityContext processUtilityContext)
+{
+	if (!EnableCreateDatabasePropagation || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	DropdbStmt *stmt = (DropdbStmt *) node;
+	char *databaseName = stmt->dbname;
+	bool missingOk = true;
+	Oid databaseOid = get_database_oid(databaseName, missingOk);
+	if (databaseOid == InvalidOid)
+	{
+		/* let regular ProcessUtility deal with IF NOT EXISTS */
+		return NIL;
+	}
+
+	ObjectAddress dbAddress = { 0 };
+	ObjectAddressSet(dbAddress, DatabaseRelationId, databaseOid);
+	if (!IsObjectDistributed(&dbAddress))
+	{
+		return NIL;
+	}
+
+	List *workerNodes = TargetWorkerSetNodeList(NON_COORDINATOR_METADATA_NODES,
+												RowShareLock);
+	if (list_length(workerNodes) == 0)
+	{
+		return NIL;
+	}
+
+	char *dropDatabaseCommand = DeparseTreeNode(node);
+
+	StringInfo internalDropCommand = makeStringInfo();
+	appendStringInfo(internalDropCommand,
+					 "SELECT pg_catalog.citus_internal_database_command(%s)",
+					 quote_literal_cstr(dropDatabaseCommand));
+
+	/* Delete from pg_dist_object */
+
+	if (IsObjectDistributed(&dbAddress))
+	{
+		UnmarkObjectDistributed(&dbAddress);
+	}
+
+	/* we execute here to avoid EnsureCoordinator check in ExecuteDistributedDDLJob */
```

Comment:
if this is the only reason, then it might be nicer to add a flag to DDLJob which controls whether to call EnsureCoordinator()

Reply:
Checked and this is not the only reason. Using ExecuteUtilityTaskList is mandatory since there are namespace checks as well.